### ISSUE TITLE
Invalid linker options when combining gcc with non-gnu linker

### ISSUE THREAD

**Issue by wouterel** [glotzerlab/hoomd-blue#1191]
## Description

Compiling of several tests fails to complete when using gcc on systems where the standard linker does not recognize the options `--allow-shlib-undefined` or `--no-as-needed`. The reason is found in lines 57-60 of `hoomd/md/test/CMakeLists.txt`,
where these linker options are added on the condition that CMAKE_CXX_COMPILER_ID equals "GNU".

This happens e.g. in MacOS when using a gcc installed through macports, which uses the system linker through xcode. I tested this in two settings (see below) and in both cases compiling failed, but commenting out lines 57-60 in the file mentioned above fixed the issue.

## Script

```bash
mkdir build
cd build
cmake -DCMAKE_C_COMPILER=$(which gcc-mp-11) -DCMAKE_CXX_COMPILER=$(which g++-mp-11) ..
make
```

## Output
CMake output for reference: [cmake_output.txt](https://github.com/glotzerlab/hoomd-blue/files/7837688/cmake_output.txt)
Only showing where it goes wrong:
```
[ 73%] Linking CXX executable test_berendsen_integrator
ld: unknown option: --allow-shlib-undefined
collect2: error: ld returned 1 exit status
make[2]: *** [hoomd/md/test/test_berendsen_integrator] Error 1
make[1]: *** [hoomd/md/test/CMakeFiles/test_berendsen_integrator.dir/all] Error 2
```
If compiling with -j8, several of these errors show up for various test exectuables.

## Expected output

Succesful compilation.

## Configuration

Platform (both giving the same result):
- Mac (Intel Mac with MacOS 10.14 and gcc 8)
- Mac (M1 Mac with MacOS 12.1 with gcc 11) (this is the system I used to generate the output above)

Installation method:
- Compiled from source

Versions:
- Python version: 3.9.9
- **HOOMD-blue** version: 3.0.0-beta12
- gcc 11.2.0
- cmake 3.22.1

## Developer
The lines the issue is about were added by @joaander but I'm also happy to contribute. I do not know what kinds of systems there are on which gnu compilers tend to be paired with non-gnu linkers. But if my understanding is correct that  one simply does not use GNU linkers in MacOS, this should be expected to happen on all MacOS systems regardless of how exactly the gnu compilers were obtained, and I would suggest to only add the linker options on non-Darwin systems by changing line 57 of the offending file to 
```
    if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" AND NOT "${CMAKE_HOST_SYSTEM_NAME}" STREQUAL "Darwin")
```
But I'd rather see someone with more CMake experience make the call on what the best solution is.

I should add that the other obvious solution (telling Mac users to use Clang instead of GCC) gave me trouble as well, but I will address that in a separate issue.

**Comment by joaander** [comment]
Yes, please contribute the proposed fix. You are in the best position to test the changes. Simplify the fix using the predefined `APPLE` variable: https://cmake.org/cmake/help/v3.22/variable/APPLE.html .

CMake doesn't provide a variable that identifies the linker (https://cmake.org/cmake/help/v3.22/manual/cmake-variables.7.html) so checking for the apple platform is the only option.

Have you tried building the head of the `master` branch? I recently fixed a compile error and a bug on Mac:
* b6748c3303daa5a39812c6ba9a824adeb2c55e13
* f986a4dacdac5b31c2f3484c875fe2e5c521cb5f

**Comment by wouterel** [comment]
Thanks. I just made the pull request.

I also tested building the head of `master`. That now compiles successfully with both clang and gcc11. The issues I mentioned earlier (which were different on the two Mac systems I mentioned) are now gone (at least on the M1 Mac, I didn't have time yet to test it on the Intel one).

I did get some warnings with clang and some notes with gcc though. But nothing code-breaking (all compilation results pass both ctest and pytest). Let me know if you would like to see the details.
